Added examples of Lists to Managing Resources (#28608)
kubernetes/website: content/en/docs/reference/using-api/api-concepts.md

You can get more information about each list type from the [Kubernetes API](https://kubernetes.io/docs/reference/kubernetes-api/) documentation.

When you query the API for a particular type, all items returned by that query are of that type. For example, when you
ask for a list of pods, the list type is shown as `kind: PodList` and each `item` in that list represents a pod. For example:

```console

GET /api/v1/namespaces/kube-system/pods
---
{
  "kind": "PodList",
  "apiVersion": "v1",
  "metadata": {
    "resourceVersion": "2873495"
  },
  "items": [
    {
      "metadata": {
        "name": "coredns-558bd4d5db-hfgxg",
        "generateName": "coredns-558bd4d5db-",
...

      "metadata": {
        "name": "weave-net-rnbqq",
        "generateName": "weave-net-",
...
```

Some tools, such as `kubectl` provide another way to query the Kubernetes API. Because the output of `kubectl` might include multiple list types, the list of items is represented as `kind: List`. For example:

```console

$ kubectl get pods -n kube-system -o yaml 

apiVersion: v1
items:
- apiVersion: v1
 kind: Pod
  metadata:

> kind: Pod

Reviewer: Is this right? The indentation looks wrong.

Reviewer: It might be easier to fetch a list of smaller objects. Maybe Lease or ServiceAccount?

Author: Good catch! Fixed. Maybe you are right. I'll try some smaller examples.

Author: @sftim I tried a few different ones and changes the examples to use Service. I showed the same result querying the api directly and through kubectl. Let me know what you think.

Note that the `resourceVersion` of the list remains constant across each request, indicating the server is showing us a consistent snapshot of the pods. Pods that are created, updated, or deleted after version `10245` would not be shown unless the user makes a list request without the `continue` token.  This allows clients to break large requests into smaller chunks and then perform a watch operation on the full set without missing any updates.

## Lists

There are dozens of list types (such as `PodList`, `ServiceList`, and `NodeList`) defined in the Kubernetes API.
You can get more information about each list type from the [Kubernetes API](https://kubernetes.io/docs/reference/kubernetes-api/) documentation.

When you query the API for a particular type, all items returned by that query are of that type. For example, when you
ask for a list of services, the list type is shown as `kind: ServiceList` and each `item` in that list represents a service. For example:

```console

> GET /api/v1/namespaces/services

Reviewer: I'd expect this to error out with `namespaces "services" not found`
Did you mean: `GET /api/v1/services`?

Author: You're right. I'll fix it in a second.

Author: Done.

---
{
  "kind": "ServiceList",
  "apiVersion": "v1",
  "metadata": {
    "resourceVersion": "2947301"
  },
  "items": [
    {
      "metadata": {
        "name": "kubernetes",
        "namespace": "default",
...
      "metadata": {
        "name": "kube-dns",
        "namespace": "kube-system",
...
```

Some tools, such as `kubectl` provide another way to query the Kubernetes API. Because the output of `kubectl` might include multiple list types, the list of items is represented as `kind: List`. For example:

```console

$ kubectl get services -A -o yaml 

apiVersion: v1
items:
- apiVersion: v1
  kind: Service
  metadata:
    creationTimestamp: "2021-06-03T14:54:12Z"
    labels:
      component: apiserver
      provider: kubernetes
    name: kubernetes
    namespace: default
...
- apiVersion: v1
  kind: Service
  metadata:
    annotations:
      prometheus.io/port: "9153"
      prometheus.io/scrape: "true"
    creationTimestamp: "2021-06-03T14:54:14Z"
    labels:
      k8s-app: kube-dns
      kubernetes.io/cluster-service: "true"
      kubernetes.io/name: CoreDNS
    name: kube-dns
    namespace: kube-system
...
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

> Keep in mind that the Kubernetes API does not have a `kind: List` type. The `kind: List` resource is more of an internal mechanism for tools such as `kubectl` to present lists of mixed resources in the same output.

Reviewer: I'd use a `{{< note >}}` callout for this. Also be consistent about whether it's a kind or a type or a resource.

Author: I'll make it an note.  As for what to call `kind: List`, I just said it wasn't a "type". And it's certainly not a Kubernetes resource. So I'll drop "The" and "resource" and just start the sentence with `kind: List`, as long as that doesn't break some docs style issue.

Author: Done.

Note that the `resourceVersion` of the list remains constant across each request, indicating the server is showing us a consistent snapshot of the pods. Pods that are created, updated, or deleted after version `10245` would not be shown unless the user makes a list request without the `continue` token.  This allows clients to break large requests into smaller chunks and then perform a watch operation on the full set without missing any updates.

## Lists

There are dozens of list types (such as `PodList`, `ServiceList`, and `NodeList`) defined in the Kubernetes API.
You can get more information about each list type from the [Kubernetes API](https://kubernetes.io/docs/reference/kubernetes-api/) documentation.

When you query the API for a particular type, all items returned by that query are of that type. For example, when you

> ask for a list of services, the list type is shown as `kind: ServiceList` and each `item` in that list represents a service. For example:

Reviewer: Writing `item` might be misleading. The `items` (plural) array contains entries that each represent a single Servcie (title case as Service is a kind in Kubernetes).

Author: Good point. I'll remove formatting from item and capitalize Service.

```console

> GET /api/v1/services

Reviewer: `kubectl --raw` might be another way to fetch this. Not sure if that helps or not.

Author: I was just trying to match the way that content is shown in the rest of the page. I usually use curl commands to query the API after running `kubectl proxy`. So I'm thinking we should leave that unless we want to change the others.

Author: Let me know if you disagree.

---
{
  "kind": "ServiceList",
  "apiVersion": "v1",
  "metadata": {
    "resourceVersion": "2947301"
  },
  "items": [
    {
      "metadata": {
        "name": "kubernetes",
        "namespace": "default",
...
      "metadata": {
        "name": "kube-dns",
        "namespace": "kube-system",
...
```

Some tools, such as `kubectl` provide another way to query the Kubernetes API. Because the output of `kubectl` might include multiple list types, the list of items is represented as `kind: List`. For example:

```console

$ kubectl get services -A -o yaml 

apiVersion: v1
items:
- apiVersion: v1
  kind: Service
  metadata:
    creationTimestamp: "2021-06-03T14:54:12Z"
    labels:
      component: apiserver
      provider: kubernetes
    name: kubernetes
    namespace: default
...
- apiVersion: v1
  kind: Service
  metadata:
    annotations:
      prometheus.io/port: "9153"
      prometheus.io/scrape: "true"
    creationTimestamp: "2021-06-03T14:54:14Z"
    labels:
      k8s-app: kube-dns
      kubernetes.io/cluster-service: "true"
      kubernetes.io/name: CoreDNS
    name: kube-dns
    namespace: kube-system
...

> kind: List

Reviewer 2: nit: move this and metadata up below `apiVersion`

Author: Okay. I moved `kind: List` and the following two lines to after `apiVersion` above.

Author: @eddiezane Please double-check that I got the example right. Thanks!

metadata:
  resourceVersion: ""
  selfLink: ""
```

{{< note >}}

> Keep in mind that the Kubernetes API does not have a `kind: List` type. `kind: List` is more of an internal mechanism for tools such as `kubectl` to present lists of mixed resources in the same output.

Reviewer 2: ```suggestion
Keep in mind that the Kubernetes API does not have a `kind: List` type. `kind: List` is an internal type for lists of mixed resources and should not be depended upon.
```

Author: That wording is fine with me. Done.
